### PR TITLE
admin, category-product listing PSR-12 formatting

### PR DESCRIPTION
**Pull request by torvista**
and removed declare(strict_types=1) from first line, which was only for testing. I leave it in all the files I mess with, seems like a good idea.
No other code changes

This was formatted by phpStorm so maybe there are things in there that are not to everyone's taste. Good to find out now.....

### CODE REVIEW

**Comment by torvista** [comment]
I don't know where I got the idea from that we are using PSR-12....docs say PSR-2. 
So, what is it and is there a problem with this loitering PR?

**Comment by lat9** [comment]
It's only been 4 days since you submitted; lack of time to review (for me anyway) is the problem.

**Comment by torvista** [comment]
Sorry, too impatient. Still, should be PSR2 or 12?

**Comment by zcwilt** [comment]
I would say PSR12, but would need to look at what we have posted previously.

On why this hasn't been merged.

The diff for this is huge and reviewing it takes some time. 
While we know you and your history, it would be easy for someone else to create a huge psr diff and then slip into that diff some 
code that compromises security.

**Comment by torvista** [comment]
In theory there is no change at all, purely formatting.

**Comment by zcwilt** [comment]
how do I k now that

**Comment by torvista** [comment]
Yes, quite, trust nothing!

File: admin/category_product_listing.php [hunk 1]
```diff
@@ -25,1230 +25,1656 @@
...
+            }
+            zen_redirect(
+                zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'cPath=' . $_GET['cPath'] . '&cID=' . $_GET['cID'] . (isset($_GET['page']) ? '&page=' . $_GET['page'] : '') . $search_parameter)
+            );
+            break;
+        case 'remove_type':
+            if (isset($_POST['type_id'])) {
+                zen_remove_restrict_sub_categories($_GET['cID'], (int)$_POST['type_id']);
+                $action = 'edit';
+                zen_redirect(zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'action=edit_category&cPath=' . $_GET['cPath'] . '&cID=' . zen_db_prepare_input($_GET['cID'])));
```
Comment (lat9) [inline admin/category_product_listing.php:121]: $_GET['cID'] is pre-sanitized to an int, no need for zen_db_prepare_input.

File: admin/category_product_listing.php [hunk 2]
```diff
@@ -25,1230 +25,1656 @@
...
-        $path = explode('_', zen_get_generated_category_path_ids($new_parent_id));
+            // future cat specific deletion
+            $delete_linked = 'true';
+            if (isset($_POST['delete_linked']) && $_POST['delete_linked'] !== '') {
+                $delete_linked = $_POST['delete_linked'];
+            }
+
+            // delete category and products
+            if (!empty($_POST['categories_id']) && $_POST['categories_id'] !== (int)TOPMOST_CATEGORY_PARENT_ID) {
+                $categories_id = zen_db_prepare_input($_POST['categories_id']);
```
Comment (lat9) [inline admin/category_product_listing.php:146]: zen_db_prepare_input not needed, variable is pre-sanitized to an int.

File: admin/category_product_listing.php [hunk 3]
```diff
@@ -25,1230 +25,1656 @@
...
+                    }
+
+                    zen_remove_category($categories[$i]['id']);
+                } // end for loop
+            }
+            zen_redirect(zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'cPath=' . $cPath));
+            break;
+        case 'move_category_confirm':
+            if (isset($_POST['categories_id']) && ($_POST['categories_id'] !== (int)$_POST['move_to_category_id'])) {
+                $categories_id = zen_db_prepare_input($_POST['categories_id']);
```
Comment (lat9) [inline admin/category_product_listing.php:211]: Pre-sanitized to an int; no need for zen_db_prepare_input.

File: admin/category_product_listing.php [hunk 4]
```diff
@@ -25,1230 +25,1656 @@
...
+                $new_parent_id = zen_db_prepare_input($_POST['move_to_category_id']);
```
Comment (lat9) [inline admin/category_product_listing.php:212]: Use the posted value cast to an int?

File: admin/category_product_listing.php [hunk 5]
```diff
@@ -25,1230 +25,1656 @@
...
-    case 'attribute_features':
-    case 'attribute_features_copy_to_product':
-    case 'attribute_features_copy_to_category':
-      break;
-    default:
-      $zco_notifier->notify('NOTIFY_ADMIN_PROD_LISTING_DEFAULT_ACTION');
-      $action = $_GET['action'] = '';
-      break;
-  }
+                        WHERE categories_id = " . (int)$categories_id
```
Comment (lat9) [inline admin/category_product_listing.php:232]: $categories_id is already 'known' to be an int.

File: admin/category_product_listing.php [hunk 6]
```diff
@@ -25,1230 +25,1656 @@
...
+                            <th class="smallText"><?php
+                                echo TEXT_LEGEND; ?></th>
+                            <th class="text-center smallText"><?php
+                                echo TEXT_LEGEND_STATUS_OFF; ?></th>
+                            <th class="text-center smallText"><?php
+                                echo TEXT_LEGEND_STATUS_ON; ?></th>
+                            <th class="text-center smallText"><?php
+                                echo TEXT_LEGEND_LINKED; ?></th>
+                            <th class="text-center smallText"><?php
+                                echo TEXT_LEGEND_META_TAGS . '<br>' . TEXT_YES . '&nbsp;' . TEXT_NO; ?></th>
```
Comment (lat9) [inline admin/category_product_listing.php:381]: Why the line-splits?  I'm open to alternate views, but I find this less readable than all-on-one-line.

File: admin/category_product_listing.php [hunk 7]
```diff
@@ -25,1230 +25,1656 @@
...
+                    if ($check_categories) {
+                        $zc_skip_products = true;
+                        $zc_skip_categories = false;
+                    }
+                    if ($check_products > 0) {
+                        $zc_skip_products = false;
+                        $zc_skip_categories = true;
+                    }
+
+                    if (isset($_GET['search']) && !empty($_GET['search'])) {
```
Comment (lat9) [inline admin/category_product_listing.php:453]: Isn't !empty() sufficient?

File: admin/category_product_listing.php [hunk 8]
```diff
@@ -25,1230 +25,1656 @@
...
+                    <?php
+                    echo zen_hide_session_id(); ?>
+                    <div class="form-group">
+                        <?php
+                        if ($_SESSION['display_categories_dropdown'] === 0) { ?>
+                            <div class="col-sm-6 col-md-4 control-label">
+                                <a href="<?php
+                                echo zen_href_link(
+                                    FILENAME_CATEGORY_PRODUCT_LISTING,
+                                    'set_display_categories_dropdown=1' . (isset($_GET['cID']) ? '&cID=' . (int)$_GET['cID'] : '') . '&cPath=' . $cPath . (isset($_GET['page']) ? '&page='
```
Comment (lat9) [inline admin/category_product_listing.php:564]: $_GET['cID'] is already sanitized to an int.

File: admin/category_product_listing.php [hunk 9]
```diff
@@ -25,1230 +25,1656 @@
...
+                                <?php
+                                echo zen_draw_pull_down_menu('cPath', zen_get_category_tree(), $current_category_id, 'onchange="this.form.submit();" class="form-control" id="cPath"'); ?>
+                            </div>
+                        <?php
+                        } else { ?>
+                            <div class="col-sm-6 col-md-4 control-label">
+                                <a href="<?php
+                                echo zen_href_link(
+                                    FILENAME_CATEGORY_PRODUCT_LISTING,
+                                    'set_display_categories_dropdown=0' . (isset($_GET['cID']) ? '&cID=' . (int)$_GET['cID'] : '') . '&cPath=' . $cPath . (isset($_GET['page']) ? '&page='
```
Comment (lat9) [inline admin/category_product_listing.php:581]: $_GET['cID'] is already sanitized to an int.

File: admin/category_product_listing.php [hunk 10]
```diff
@@ -25,1230 +25,1656 @@
...
+                        <th><?php
+                            echo TABLE_HEADING_CATEGORIES_PRODUCTS; ?></th>
+                        <th class="hidden-sm hidden-xs"><?php
+                            echo TABLE_HEADING_IMAGE; ?></th>
+                        <?php
+                        if ($show_prod_labels) { ?>
+                            <th class="hidden-sm hidden-xs"><?php
+                                echo TABLE_HEADING_MODEL; ?></th>
+                            <th class="text-right hidden-sm hidden-xs"><?php
+                                echo TABLE_HEADING_PRODUCTS_PRICE; ?></th>
```
Comment (lat9) [inline admin/category_product_listing.php:663]: As above, I find splitting these <th> tags onto multiple lines to less readable.

File: admin/category_product_listing.php [hunk 11]
```diff
@@ -25,1230 +25,1656 @@
...
+                        <?php
+                        } ?>
+                        <?php
```
Comment (lat9) [inline admin/category_product_listing.php:666]: Can't we just remove this ?> <?php pair?

File: admin/category_product_listing.php [hunk 12]
```diff
@@ -25,1230 +25,1656 @@
...
+                            }
+                        }
+                        ?>
+                        <?php
+                        if ($show_prod_labels || SHOW_COUNTS_ADMIN === 'true') { ?>
+                            <th class="text-right hidden-sm hidden-xs"><?php
+                                echo TABLE_HEADING_QUANTITY; ?></th>
+                        <?php
+                        } ?>
+                        <?php
```
Comment (lat9) [inline admin/category_product_listing.php:706]: Remove the ?> <?php, replacing with a blank line.

File: admin/category_product_listing.php [hunk 13]
```diff
@@ -25,1230 +25,1656 @@
...
+                        ?>
+                        <th class="text-right"><?php
+                            echo TABLE_HEADING_STATUS; ?></th>
+                        <?php
+                        if ($action === '') {
+                            ?>
+                            <th class="text-right hidden-sm hidden-xs"><?php
+                                echo TABLE_HEADING_CATEGORIES_SORT_ORDER; ?></th>
+                            <th class="text-right"><?php
+                                echo TABLE_HEADING_ACTION; ?></th>
```
Comment (lat9) [inline admin/category_product_listing.php:748]: As before, use a single line for each <th>

File: admin/category_product_listing.php [hunk 14]
```diff
@@ -25,1230 +25,1656 @@
...
-                  <td class="hidden-sm hidden-xs"><!-- no price for categories --></td>
-                <?php } ?>
-                <?php if (SHOW_COUNTS_ADMIN === 'true') { ?>
-                  <td class="text-right hidden-sm hidden-xs">
+                        if (isset($_GET['search'])) {
+                            $cPath = $category['parent_id'];
+                        }
+
+                        if ((!isset($_GET['cID']) && !isset($_GET['pID']) || (isset($_GET['cID']) && ($_GET['cID'] == $category['categories_id']))) && !isset($cInfo)
+                            && (substr($action, 0, 3) != 'new')
```
Comment (lat9) [inline admin/category_product_listing.php:763]: Use `strpos($action, 'new') !== 0`

File: admin/category_product_listing.php [hunk 15]
```diff
@@ -25,1230 +25,1656 @@
...
+                    // Split Page
+
+                    foreach ($products as $product) {
+                        $products_count++;
+// Get categories_id for product if search
+                        if (isset($_GET['search'])) {
+                            $cPath = $product['master_categories_id'];
+                        }
+
+                        if ((!isset($_GET['pID']) && !isset($_GET['cID']) || (isset($_GET['pID']) && ($_GET['pID'] === $product['products_id']))) && !isset($pInfo) && !isset($cInfo)
```
Comment (lat9) [inline admin/category_product_listing.php:1005]: $_GET['pID'] is sanitized to an int, while the database return of the 'products_id' is a string.

**Review by lat9 — CHANGES_REQUESTED** [review]
You asked for it!

File: admin/category_product_listing.php [hunk 16]
```diff
@@ -25,1230 +25,1656 @@
...
-  $_GET['page'] = (int)$_GET['page'];
+    $_GET['page'] = (int)$_GET['page'];
 }
 if (isset($_GET['product_type'])) {
-  $_GET['product_type'] = (int)$_GET['product_type'];
+    $_GET['product_type'] = (int)$_GET['product_type'];
 }
 if (isset($_GET['cID'])) {
-  $_GET['cID'] = (int)$_GET['cID'];
+    $_GET['cID'] = (int)$_GET['cID'];
```
[the remark this answers is not in the document]
Comment (torvista) [reply admin/category_product_listing.php:37]: ok, but this change is I think outside of this formatting-only PR/for later.

File: admin/category_product_listing.php [hunk 17]
```diff
@@ -25,1230 +25,1656 @@
...
-              $do_delete_flag = true;
-            }
+    switch ($action) {
+        case 'set_categories_products_sort_order':
+            $_SESSION['categories_products_sort_order'] = $_GET['reset_categories_products_sort_order'];
+            $action = '';
+            zen_redirect(
+                zen_href_link(
+                    FILENAME_CATEGORY_PRODUCT_LISTING,
+                    'cPath=' . $_GET['cPath'] . ((isset($_GET['pID']) && !empty($_GET['pID'])) ? '&pID=' . $_GET['pID'] : '') . $search_parameter . ((isset($_GET['page']) && !empty($_GET['page']))
```
[the remark this answers is not in the document]
Comment (torvista) [reply admin/category_product_listing.php:56]: Correct. For next PR.

File: admin/category_product_listing.php [hunk 18]
```diff
@@ -25,1230 +25,1656 @@
...
+                )
+            );
+            break;
+        case 'set_editor':
+            // Reset will be done by init_html_editor.php. Now we simply redirect to refresh page properly.
+            $action = '';
+            zen_redirect(
+                zen_href_link(
+                    FILENAME_CATEGORY_PRODUCT_LISTING,
+                    'cPath=' . $_GET['cPath'] . ((isset($_GET['pID']) && !empty($_GET['pID'])) ? '&pID=' . $_GET['pID'] : '') . ((isset($_GET['page']) && !empty($_GET['page'])) ? '&page='
```
[the remark this answers is not in the document]
Comment (torvista) [reply admin/category_product_listing.php:67]: ok, for next PR.

File: admin/category_product_listing.php (hunk 1 above)
Comment (torvista) [reply admin/category_product_listing.php:121]: ok, for next PR.

File: admin/category_product_listing.php (hunk 3 above)
Comment (torvista) [reply admin/category_product_listing.php:211]: ok, for next PR.

File: admin/category_product_listing.php (hunk 5 above)
Comment (torvista) [reply admin/category_product_listing.php:232]: I seem to recall a discussion about these unnecessary castings in sql being retained as a bit of a safety belt-and-braces approach. Usually the IDE points out these surplus castings: comes under Remove surplus sanitation: https://github.com/zencart/documentation/issues/879

For next PR.

File: admin/category_product_listing.php (hunk 6 above)
Comment (torvista) [reply admin/category_product_listing.php:381]: I don't like this either. This entire file was an auto-format deliberately to see how we like the results.

File: admin/category_product_listing.php (hunk 7 above)
Comment (torvista) [reply admin/category_product_listing.php:453]: Correct.
Next PR.

File: admin/category_product_listing.php (hunk 9 above)
Comment (torvista) [reply admin/category_product_listing.php:581]: Correct. Next PR.

File: admin/category_product_listing.php (hunk 10 above)
Comment (torvista) [reply admin/category_product_listing.php:663]: Agreed.

File: admin/category_product_listing.php (hunk 13 above)
Comment (torvista) [reply admin/category_product_listing.php:748]: Agreed.

File: admin/category_product_listing.php (hunk 14 above)
Comment (torvista) [reply admin/category_product_listing.php:763]: ok, next PR.

File: admin/category_product_listing.php [hunk 19]
```diff
@@ -25,1230 +25,1656 @@
...
+                        . $_GET['page'] : '')
+                )
+            );
+            break;
+        case 'update_category_status':
+            // disable category and products including subcategories
+            if (isset($_POST['categories_id'])) {
+                $categories_id = zen_db_prepare_input($_POST['categories_id']);
```
[the remark this answers is not in the document]
Comment (torvista) [reply admin/category_product_listing.php:75]: ok, for next PR

File: admin/category_product_listing.php [hunk 20]
```diff
@@ -25,1230 +25,1656 @@
...
+                    $products_status = isset($_POST['set_products_status']) && $_POST['set_products_status'] === 'set_products_status_on' ? '1' : ''; //Disable products or no change?
+                }
 
-            if (!empty($cascaded_prod_id_for_delete) && !empty($cascaded_prod_cat_for_delete)) {
-              $product_id = $cascaded_prod_id_for_delete;
-              $product_categories = $cascaded_prod_cat_for_delete;
-              $do_delete_flag = true;
+                for ($i = 0, $n = count($categories); $i < $n; $i++) {
+                    //set categories_status
+                    if ($categories[$i]['id'] === $categories_id) {//always update THIS category
```
[the remark this answers is not in the document]
Comment (torvista) [reply admin/category_product_listing.php:95]: This has a gotcha.
$categories_id is ini-cast to integer.
$categories is from zen_category_tree, comes from a query and so the 'id' is a string....except for the first element, which is the passed parameter $categories_id and is an integer.
So, in the function zen_category_tree, this passed parameter should be cast to string: another PR.

Conclusion, this line is a bug: another PR.

File: admin/category_product_listing.php (hunk 2 above)
Comment (torvista) [reply admin/category_product_listing.php:146]: ok, next PR.

File: admin/category_product_listing.php (hunk 8 above)
Comment (torvista) [reply admin/category_product_listing.php:564]: ok, next PR

File: admin/category_product_listing.php (hunk 4 above)
Comment (torvista) [reply admin/category_product_listing.php:212]: You mean 'move_to_category_id' should be added to the init-sanitize cast to int group?

**Comment by proseLA** [comment]
> 
> This was formatted by phpStorm so maybe there are things in there that are not to everyone's taste. Good to find out now.....

this is such a misleading statement.  phpStorm only does what you tell it.  the settings for code style are yours and yours alone.  you re-formatted this file.  do you like it?  can you replicate so that everyone else can reformat it as such?

i will say that these files with a combination of html and php are always the trickiest to format.

i have played around with various reformatting tools using your file and the base file as well.  all while trying to stay within the PSR-12 formatting rules.  and frankly there is no consistency.  and until someone comes out with what our definitive formatting rules are; i'm just not sure what to say....  i'll give you my opinion on this file and these changes.  don't like it.  at all.

the splitting onto multiple lines in numerous places is a choice that you made.  and unless you can show me how that is required in PSR12, i am not a fan.

i still say replacing `<?php echo ` with `<?= ` in template files with html make code much more legible.  at some point, i am confident that the maintainers will agree with me.  despite the resistance.  change is hard; but this one stands out to me.  

https://gist.github.com/ryansechrest/8138375#5-short-echo-tag
https://github.com/WordPress/WordPress-Coding-Standards/issues/1642

the comma after the last array element is something the maintainers have approved.  see:

https://docs.zen-cart.com/dev/contributing/coding_standards/#arrays

it is in our coding standards.  and yet, you have violations all over this file.  did phpstorm do it?  did you do it?

image

**Comment by torvista** [comment]
> can you replicate so that everyone else can reformat it as such?

That was part of the idea.
This is really a test case. 
There are a mass of options for formatting in phpStorm. But, you can "Set from" a reference. So I used the PSR12 reference with no further modifications precisely to see if that would be acceptable and so easily replicated. So it is not my formatting at all.

So far I've made two modifications to that PSR-12: one to prevent the wrapping after the ?php and another to not put each function call arguments on a new line.
It's a big file and so should provide examples of formatting all types of code structure. 

Adding commas after arrays is not something I was aware of as being a suggested code change to do by default.
There are no code changes in this PR but plenty of suggestions for the next.

`i still say replacing <?php echo with <?= in template files with html make code much more legible`
I agree.

**Comment by torvista** [comment]
`i have played around with various reformatting tools`
Is there one you like?

I've looked at this again and find that PSR2 is much less aggressive than PSR12 and more resembles the original code apart from the identing.

In both cases the uglier results are due to applying the 80 char wrapping.

Here are the files for easier comparison
[compare formats.zip](https://github.com/zencart/zencart/files/9331035/compare.formats.zip)

Personally I would format a file to PSR2 before touching it, then while editing one could improve any ugly bits.

File: admin/category_product_listing.php (hunk 15 above)
Comment (torvista) [reply admin/category_product_listing.php:1005]: Prior to correcting this, perhaps this clause needs a review.
I'm not sure what cID/$cInfo has to do with this.

Putting it In a more readable form:
```
if (
                    (
                        !isset($_GET['pID']) 
                        && !isset($_GET['cID']) 
                        || (isset($_GET['pID']) && ($_GET['pID'] === $product['products_id']))
                    )
                    && !isset($pInfo)
                    && !isset($cInfo)
                    && (strpos($action,'new') !== 0))
   {
      $pInfo = new objectInfo($product);
    }
```

Comment (scottcwilson) [reply admin/category_product_listing.php:1005]: You have to handle these long lines yourself or you'll hit a hard break.  Something like this:

```
                        if ((!isset($_GET['pID']) && !isset($_GET['cID']) || 
                                (isset($_GET['pID']) && ($_GET['pID'] === $product['products_id']))) 
                            && !isset($pInfo) && !isset($cInfo) && (strpos($action,'new') !== 0)) {
```

**Comment by scottcwilson** [comment]
Please defer reformatting changes until we get a 1.5.8 release and we come to a consensus about how we're going to do this.

**Pull request closed without merging**